File: test-errors/src/programmers/helpers/OptionPredicator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OptionPredicator = void 0;
var OptionPredicator;
(function (OptionPredicator) {
    OptionPredicator.numeric = (options) => OptionPredicator.finite(options) || options.numeric === true;
    OptionPredicator.functional = (options) => options.functional === true;
    OptionPredicator.finite = (options) => options.finite === true;
    OptionPredicator.undefined = (options) => options.undefined !== false;
})(OptionPredicator || (exports.OptionPredicator = OptionPredicator = {}));

File: packages/typia/native/core/programmers/helpers/OptionPredicator.go
package helpers

import (
  nativecontext "github.com/samchon/typia/packages/typia/native/core/context"
  schemametadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"
)

type optionPredicatorNamespace struct{}

var OptionPredicator optionPredicatorNamespace

func (optionPredicatorNamespace) Numeric(options nativecontext.ITransformOptions) bool {
  return OptionPredicator.Finite(options) || options.Numeric != nil && *options.Numeric
}

func (optionPredicatorNamespace) Functional(options nativecontext.ITransformOptions) bool {
  return options.Functional != nil && *options.Functional
}

func (optionPredicatorNamespace) Finite(options nativecontext.ITransformOptions) bool {
  return options.Finite != nil && *options.Finite
}

func (optionPredicatorNamespace) Undefined(options nativecontext.ITransformOptions) bool {
  return options.Undefined == nil || *options.Undefined
}

func (optionPredicatorNamespace) StrictOptionalUndefined(context nativecontext.ITypiaContext, metadata *schemametadata.MetadataSchema) bool {
  return OptionPredicator.ExactOptionalProperty(context, metadata) &&
    metadata.Required == false
}

func (optionPredicatorNamespace) ExactOptionalProperty(context nativecontext.ITypiaContext, metadata *schemametadata.MetadataSchema) bool {
  return context.CompilerOptions != nil &&
    context.CompilerOptions.ExactOptionalPropertyTypes.IsTrue() &&
    OptionPredicator.Undefined(context.Options) == false &&
    metadata != nil &&
    metadata.Optional
}
